Level 3

MDaryaK opened the pull request.

MDaryaK committed:
wip

MDaryaK committed:
fixes

MDaryaK committed:
wip

MDaryaK committed:
Update readme.md

MDaryaK committed:
profile ready

irinamozes commented:
Здравствуйте, Дарья.

По предоставленной Вами ссылке https://movies-darya.nomoredomainsrocks.ru открывается пустая страница с ошибкой: "500 Internal Server Error" (см. снимок). 
![Снимок экрана в 2023-11-01 17-38-52](https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/256e9247-9305-4aa8-9822-2bda1baeced8)
Разберитесь, пожалуйста, в чём дело, и присылайте диплом снова.
Текущая итерация засчитываться не будет.

MDaryaK committed:
wip

MDaryaK committed:
fixes

MDaryaK committed:
fixes

MDaryaK committed:
eslint fix

MDaryaK committed:
wip

MDaryaK committed:
Update index.js

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    **Комментарий: при выходе авторизованного пользователя на главную страницу по ссылке-логотипу или по URL-адресу  этой страницы в адресной строке браузера шапка на главной странице содержит пункты меню неавторизованного пользователя, а должна содержать пункты меню «Фильмы», «Сохранённые фильмы» и «Аккаунт».**
    
      **Комментарий: при нажатии кнопки «Назад» на странице "404" авторизованный пользователь попадает на главную страницу, и шапка на главной странице содержит пункты меню неавторизованного пользователя, а должна содержать пункты меню «Фильмы», «Сохранённые фильмы» и «Аккаунт».**
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
   **Комментарий: повторный показ последнего поиска пользователя при повторном входе на страницу фильмов, о чём идёт речь в текущем требовании чек-листа, не реализован.**
   
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит сайт, или перейдёт на страницу фильмов с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска, который был перед перезагрузкой или уходом на другую страницу (с расстановкой актуального состояния лайков). При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать. Сейчас пользователь не видит результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта.**     
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: При вводе в поле email адреса почты без точки перед именем домена верхнего уровня (в правой части адреса) можно выполнить запрос на сервер, что противоречит текущему требованию чек-листа, так как в этом случае адрес будет невалидным. Нужно, чтобы валидация имени, почты и пароля и на серверной и на клиентской частях приложения, была одинаковой (по одним и тем же регулярным выражениям), и при невалидных данных запрос на сервер вообще не возможно было отправить.**

    **Комментарий: Еесли на клиенте происходит недостаточно строгая валидация адреса email, и на сервер можно отослать невалидный адрес с точки зрения сервера, нужно подобрать регулярное выражение, рекомендуемое в интернете для проверки email, чтобы строгость проверки email на клиенте, совпадала со строгостью проверки email на сервере. Тогда при запросах на регистрацию или авторизацию, сервер не будет присылать в ответе ошибку 400, признавая, что адрес почты валиден. В этом и состоит смысл текущего пункта чек-листа.**
    
   **Комментарий: запрос на сервер можно выполнить даже при незаполненном поле (см. снимок), что также ⚠️противоречит текущему требованию чек-листа.**
![Снимок экрана в 2023-11-02 16-25-01](https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/7c55cac3-cbb7-41bb-94f5-fda8fa05bd22) 
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    **Комментарий: Если запрос прошёл успешно, то происходит редирект на страницу авторизации, а нужен автоматический вход на `/movies` как авторизованного пользователя, без необходимости посещения страницы авторизации.**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
  **Комментарий: в приветствии отображается дефолтное имя, а не имя пользователя.**
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована другим пользователем, сервер приложения должен присылать ошибку 409 в ответе.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    **Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
	**Комментарий: При входе на страницу профиля и нажатии кнопки "Редактировать" появляется кнопка «Сохранить» в активном состоянии (см. снимок). И пользователь может сохранить текущие данные профиля, не изменяя их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в стейте и в контексте.**	
![Снимок экрана в 2023-11-02 15-31-49](https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/fdf485fb-2087-48f0-aa28-e351bb6779af)

	**Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
	
	**Напоминание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  **Комментарий: Вывод карточек на экран не соответствует постановке задачи, данной в описании диплома, хотя бы потому, что на экран всегда должны выводиться одинаковые по количеству карточек ряды карточек ( возможно, кроме последнего ряда), а сейчас это не так (см. снимок при ширине экрана 1220px).**
![Снимок экрана в 2023-11-02 15-52-08](https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/8276e0ee-799d-4ef9-bbe7-771010308486)

 **Ориентироваться нужно на описание функционала рендера карточек, которое дано в задании по диплому:**
   
> Обратите внимание, что количество карточек, которые отображаются на странице, зависит от ширины экрана устройства.
>     • Ширина 1280px — 4 ряда карточек. Кнопка «Ещё» загружает дополнительный ряд карточек.
>     • Ширина 768px — 4 ряда карточек. Кнопка «Ещё» загружает ряд карточек.
>     • Ширина от 320px до 480px — 5 карточек по 1 в ряд. Кнопка «Ещё» загружает по 2 карточки.
> В ряду отображайте столько карточек, сколько отображается в вашем макете на соответствующей ширине экрана.

 **Комментарий (пояснение, как должен быть реализован функционал вывода карточек, чтобы удовлетворить требованиям, приведённым в цитате выше):**  
  **1. При изменении пользователем ширины экрана результат поиска нужно выводить с самого начала с применением дебаунса. То есть, когда пользователь остановится на какой-то ширине экрана, результат нужно выводить снова, начиная с первых 4-х или 5-ти рядов, после которых вывести кнопку "Ещё". Результат выводить в соответствии с правилами для текущей ширины экрана.**
  
  **2. При осуществлении нового поиска на той же ширине экрана, при которой был предыдущий поиск, вывод карточек нужно начинать с самого начала, то есть с первых 4-х (или 5-ти) рядов, после которых нужно вывести кнопку "Ещё".**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
    **Комментарий: При клике на неактивную кнопку «Лайк» в блоке карточки на странице фильмов, кнопка не приобретает активный статус, карточка в избранные не добавляется. Возможно, запрос происходит очень медленно, в течении нескольких минут, что также не приемлемо. Карточка должна добавляться практически мгновенно и кнопка лайка приобретать активный статус.**
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [ ] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: Все исходные числовые данные проекта - длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные, значения которым задаются до начала выполнения приложения, нужно выносить в отдельный файл констант, чтобы эти исходные данные было легко менять. В коде принято использовать только имена этих констант, чтобы не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов:
**Из-за того, что не выполняется один из основных обязательных функционалов приложения - добавление карточек в избранные, на данный момент работу невозможно оценить в баллах. Нужно исправить все замечания из раздела "Функциональность проекта из брифа реализована полностью", протестировать и отладить проект перед отправкой на следующее ревью.**

**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет.**

MDaryaK committed:
fix empty signup field

Комментарий: запрос на сервер можно выполнить даже при незаполненном поле (см. снимок), что также не противоречит текущему требованию чек-листа.

MDaryaK committed:
Update index.js

MDaryaK committed:
fix header and 404 back

Комментарий: при нажатии кнопки «Назад» на странице "404" авторизованный пользователь попадает на главную страницу, и шапка на главной странице содержит пункты меню неавторизованного пользователя, а должна содержать пункты меню «Фильмы», «Сохранённые фильмы» и «Аккаунт».

Комментарий: при выходе авторизованного пользователя на главную страницу по ссылке-логотипу или по URL-адресу этой страницы в адресной строке браузера шапка на главной странице содержит пункты меню неавторизованного пользователя, а должна содержать пункты меню «Фильмы», «Сохранённые фильмы» и «Аккаунт».

irinamozes commented:
@MDaryaK
Здравствуйте!
Я сняла для Вас видео. На нём видно, что сразу при клике кнопка лайка у меня не становится активной, и карточка не появляется на странице сохранённых фильмов. На этом же видео обнаружилось, что кнопки
 стали активными и карточки появились в избранных после посещения страницы профиля. Это произошло, видимо потому, что при входе на страницу профиля произошла перезагрузка сайта. То есть карточки в избранных появляются только после перезагрузки сайта, что, конечно, не верно, всё должно происходить сразу.
https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/fed4ca4b-e32d-44ef-b149-a444749e4101

Также, обратите внимание, что при входе на любые страницы приложения при клике на пункты меню в шапке сайта никаких перезагрузок быть не должно. Сайт перезагружается при входе на страницу профиля, видимо, из -за того, что Вы используете для ссылки на страницу тег "a", а не компоненту Link. Нужно использовать Link.

Также пользователь, с которым было видео - review42@gmail.com, пароль 12345678, вышел из аккаунта. А затем авторизовался на сайте вновь.  При входе этого пользователя на сайт после логаута, на странице фильмов обнаружилось, что кнопки лайка на сохранённых карточках опять отображаются в неактивном статусе, а на странице «Сохранённые фильмы» не было никаких карточек. Но, расстановка статусов лайков и показ сохранённых карточек в любой момент нахождения пользователя на сайте должны быть актуальными и соответствовать идентификатору текущего пользователя. Статусы лайков и карточки стали отображаться правильно опять же только после перезагрузки сайта.

Также вместо комментария

> Комментарий: запрос на сервер можно выполнить даже при незаполненном поле (см. снимок), что также не противоречит текущему требованию чек-листа.

Должен быть комментарий:

"**Комментарий: запрос на сервер можно выполнить даже при незаполненном поле (см. снимок), что также ⚠️противоречит текущему требованию чек-листа.**"

Извините.

 @MDaryaK 
Ещё подсказка:
**Нужно учитывать, как ведут себя стейт-переменные, возвращаемые из useState, когда сайт перезагружается пользователями, и когда сайт работает без перезагрузок.**

MDaryaK committed:
fix 'a' links, fix eslint, fix favorite on login

Комментарий: При клике на неактивную кнопку «Лайк» в блоке карточки на странице фильмов, кнопка не приобретает активный статус, карточка в избранные не добавляется. Возможно, запрос происходит очень медленно, в течении нескольких минут, что также не приемлемо. Карточка должна добавляться практически мгновенно и кнопка лайка приобретать активный статус.

MDaryaK committed:
fix page on resize and filter

При изменении пользователем ширины экрана результат поиска нужно выводить с самого начала с применением дебаунса. То есть, когда пользователь остановится на какой-то ширине экрана, результат нужно выводить снова, начиная с первых 4-х или 5-ти рядов, после которых вывести кнопку "Ещё". Результат выводить в соответствии с правилами для текущей ширины экрана.

MDaryaK committed:
fix signup login and click on favorite icon

На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут /signup, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу /movies.

MDaryaK committed:
contstants

MDaryaK committed:
fix columns

На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.

MDaryaK committed:
fix email regex

MDaryaK committed:
save search to localstorage

MDaryaK committed:
fix eslint

MDaryaK committed:
Update index.js

MDaryaK committed:
Update index.js

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Сохранять данные  поиска на странице «Сохранённые фильмы» в localstorage не нужно, так как нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои избранные карточки. Сейчас эти данные заносятся в localstorage в те же переменные checked и search, что и данные со страницы фильмов, заменяя их значения, что может привести к неправильному повторному показу результата поиска на странице фильмов. Также при переходе со страницы на страницу, значения в переменных checked и search приводятся к дефолтным, и на странице фильмов, при переходе на неё, показываются все карточки, а не результат последнего поиска. Но, localstorage для того и нужна, чтобы в ней сохранялись данные при любых переходах и перезагрузках. Функциолнал необходимо протестировать, можно на пользователе review1@gmail.com, пароль 12345678, у которого есть сохранённые карточки со слогом "ро" в названии.**
    
    **Комментарий (очень важное замечание для правильного выполнения требования текущего пункта чек-листа): Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью независимы друг от друга.**

    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит сайт, или перейдёт на страницу фильмов с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска, который был перед перезагрузкой или уходом на другую страницу (с расстановкой актуального состояния лайков). При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.**     
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными. 
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
  **Комментарий: в приветствии отображается дефолтное имя, а не имя пользователя.**
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована другим пользователем, сервер приложения должен присылать ошибку 409 в ответе.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    **Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  **Комментарий: При нажатии кнопки "Ещё" добавляется не один или 2 ряда (когда карточки рендерятся по одной в ряд), а четыре ряда карточек, что не соответствует заданию.**

**Ориентироваться нужно на описание функционала рендера карточек, которое дано в задании по диплому:**

>   Обратите внимание, что количество карточек, которые отображаются на странице, зависит от ширины экрана устройства.
>     • Ширина 1280px — 4 ряда карточек. Кнопка «Ещё» загружает дополнительный ряд карточек.
>     • Ширина 768px — 4 ряда карточек. Кнопка «Ещё» загружает ряд карточек.
>     • Ширина от 320px до 480px — 5 карточек по 1 в ряд. Кнопка «Ещё» загружает по 2 карточки.
> В ряду отображайте столько карточек, сколько отображается в вашем макете на соответствующей ширине экрана.
> 

  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 65.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет.**

MDaryaK committed:
fix search

Комментарий (очень важное замечание для правильного выполнения требования текущего пункта чек-листа): Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью независимы друг от друга.

MDaryaK committed:
fix username

Комментарий: в приветствии отображается дефолтное имя, а не имя пользователя.

MDaryaK committed:
fix more films button

Комментарий: При нажатии кнопки "Ещё" добавляется не один или 2 ряда (когда карточки рендерятся по одной в ряд), а четыре ряда карточек, что не соответствует заданию.

MDaryaK committed:
add remove search data on logout

MDaryaK committed:
add text on profile save

MDaryaK committed:
merge

MDaryaK committed:
Update index.js

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.     
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными. 
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
      **Комментарий: При каждом вводе символа пользователем при редактировании значения поля ввода, курсор уходит в конец введённого значения (см. снимок, Яндекс Браузер), что неудобно для пользователя. Курсор должен оставаться там, где был ввод.**
![Снимок экрана в 2023-11-04 10-29-29](https://github.com/MDaryaK/movies-explorer-frontend/assets/19558891/8a5b66f1-54e8-4a47-9454-cee5d52c1bfb)  
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Как упоминалось при прошлой проверке, если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а не результат какого-либо поиска.**
## Количество баллов: 96.
4 балла снято за замечание из раздела " Доп. Комментарии".
## У Вас очень хороший дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Основы TypeScript
https://code-basics.com/ru/languages/typescript

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Борис Черный
Профессиональный TypeScript. Разработка масштабируемых JavaScript-приложений

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
